### ISSUE TITLE
現状のアーキテクチャや実装の問題点の改善

### ISSUE THREAD

**Issue by rsasaki0109** [rsasaki0109/graphslam_ros2#126]
## 概要

lidarslam_ros2の現状のコード品質とアーキテクチャの問題点を改善する。

---

### 1. 単一責任原則の大規模違反

**ScanMatcherComponent (594行)** が7つの責務を担当：
- スキャンマッチング
- LiDAR歪み補正（IMU統合）
- マップ管理
- ポーズ推定
- TF管理
- パス追跡
- パラメータ管理

**問題点：**
- 30個以上のメンバ変数
- 20個以上のパラメータ
- 循環的複雑度 ~20（推奨値: <10）

**改善案：**
```
ScanMatcherComponent
├── RegistrationStrategy (NDT/GICP切り替え)
├── MapManager (マップ管理)
├── PoseEstimator (ポーズ推定・TF)
├── LidarUndistortion (歪み補正)
└── ParameterManager (パラメータ管理)
```

### 2. スレッドセーフティの欠如

```cpp
// 危険なパターン
mapping_thread_.detach();  // スレッド終了を追跡していない
```

**問題点：**
- `targeted_cloud_` への並行アクセスが保護されていない
- `LidarUndistortion` 内の20個の配列が非スレッドセーフ
- IMUコールバックとクラウドコールバックの競合状態
- デストラクタでスレッド終了を待機していない（UBリスク）

**改善案：**
- `detach()` → `join()` に変更
- データ共有を `std::atomic` や thread-safe queue で保護
- RAII パターンでスレッド管理

### 3. 密結合したアルゴリズム選択

```cpp
// 現状（問題あり）
if (registration_method_ == "NDT") { ... }
else if (registration_method_ == "GICP") { ... }
else { exit(1); }
```

**改善案：**
```cpp
// Strategy + Factory パターン
class RegistrationStrategy {
public:
    virtual ~RegistrationStrategy() = default;
    virtual Eigen::Matrix4f align(const Cloud& source, const Cloud& target) = 0;
};

registration_ = RegistrationFactory::create(registration_method_, params);
```

### 4. 重い処理がメインスレッドをブロック

```cpp
// graph_based_slam_component.cpp
void doPoseAdjustment(...) {
    optimizer.optimize(10);  // 数秒ブロック
    pcl::io::savePCDFileASCII("map.pcd", *map_ptr);  // I/O待機
}
```

**改善案：**
- 重い処理を別スレッド/ノードに分離
- 非同期処理 + 結果のコールバック通知

### 5. 状態管理の複雑さ

```cpp
// 9個のフラグが分散
bool initial_pose_received_, initial_cloud_received_, mapping_flag_,
     is_map_updated_, set_initial_pose_, publish_tf_, use_odom_, use_imu_, debug_flag_;
```

**改善案：**
```cpp
enum class ScanMatcherState {
    Uninitialized,
    WaitingForPose,
    WaitingForCloud,
    Running,
    Mapping,
    Error
};
```

### 6. テスタビリティゼロ

**改善案：**
- インターフェース抽出 + Dependency Injection
- TF操作、アルゴリズム、センサー入力をインターフェース化

---

### 高優先度

- [ ] コンパイラ警告が無効化されている
- [ ] `std::cout`によるロギング（58箇所）→ RCLCPP_*に置換
- [ ] `exit(1)`による強制終了（2箇所）→ 例外処理に変更
- [ ] タイポ: `num_adjacent_pose_cnstraints`, `num_tareged_cloud`

### 中優先度

- [ ] C++14 → C++17への移行
- [ ] `new`演算子（58箇所）→ `std::make_shared`に置換
- [ ] `std::lock_guard` → `std::scoped_lock`に置換
- [ ] マジックナンバーの定数化

### 低優先度

- [ ] NDT/GICP初期化コードの重複除去
- [ ] パラメータバリデーションの追加
- [ ] フレームIDのパラメータ化
- [ ] ユニットテストの追加

---

## 定量的分析

| メトリクス | ScanMatcher | GraphBasedSlam | 推奨値 |
|-----------|-------------|----------------|--------|
| 行数 | 594 | 376 | < 300 |
| メソッド数 | 9 | 4 | < 5 |
| メンバ変数数 | 30+ | 15+ | < 10 |
| 循環的複雑度 | ~20 | ~15 | < 10 |
| パラメータ数 | 20+ | 7 | < 5 |
| 責務数 | 7+ | 3 | 1 |

---

## 改善の優先順位

| 順位 | 問題 | 影響 | 工数 |
|------|------|------|------|
| 1 | スレッドセーフティ | 実行時エラー・UB | 中 |
| 2 | 責務分離 | 保守性・拡張性 | 大 |
| 3 | アルゴリズム戦略化 | 拡張性 | 中 |
| 4 | 状態管理（FSM） | バグ防止 | 小 |
| 5 | 重い処理の分離 | パフォーマンス | 中 |
| 6 | テスタビリティ | 品質保証 | 大 |
